libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,880 @@
+         }
+
+         bool operator>=(const iterator& b) const {
+            return compare(b) >= 0;
+         }
+
+      private:
+         eosio::name contract_name;
+         const kv_index* idx;
+
+         uint32_t itr;
+         kv_it_stat itr_stat;
+
+         key_type key() const {
+            return idx->get_key(value());
+         }
+
+         int compare(const iterator& b) const {
+            bool a_is_end = !itr || itr_stat == kv_it_stat::iterator_end;
+            bool b_is_end = !b.itr || b.itr_stat == kv_it_stat::iterator_end;
+            if (a_is_end && b_is_end) {
+               return 0;
+            } else if (a_is_end && b.itr) {
+               return 1;
+            } else if (itr && b_is_end) {
+               return -1;
+            } else {
+               return internal_use_do_not_use::kv_it_compare(itr, b.itr);
+            }
+         }
+
+         friend kv_index;
+      };
+
+   public:
+      eosio::name name{0};
+      eosio::name table_name;
+      eosio::name contract_name;
+
+      kv_table* tbl;
+
+      kv_index() = default;
+
+      template <typename KF>
+      kv_index(KF&& kf) {
+         key_function = [=](const T& t) {
+            return make_key(std::invoke(kf, &t));
+         };
+      }
+
+      template <typename KF>
+      kv_index(eosio::name name, KF&& kf) : name{name} {
+         key_function = [=](const T& t) {
+            return make_key(std::invoke(kf, &t));
+         };
+      }
+
+      /**
+       * Search for an existing object in a table by the index, using the given key.
+       * @ingroup keyvalue
+       *
+       * @tparam K - The type of the key. This will be auto-deduced by the key param.
+       *
+       * @param key - The key to search for.
+       * @return An iterator to the found object OR the `end` iterator if the given key was not found.
+       */
+      template <typename K>
+      iterator find(K&& key) {
+         auto t_key = table_key(prefix, make_key(std::forward<K>(key)));
+
+         uint32_t itr = internal_use_do_not_use::kv_it_create(tbl->db_name, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, t_key.data(), t_key.size());
+
+         auto cmp = internal_use_do_not_use::kv_it_key_compare(itr, t_key.data(), t_key.size());
+
+         if (cmp != 0) {
+            internal_use_do_not_use::kv_it_destroy(itr);
+            return end();
+         }
+
+         return {contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+      }
+
+      /**
+       * Get the value for an existing object in a table by the index, using the given key.
+       * @ingroup keyvalue
+       *
+       * @tparam K - The type of the key. This will be auto-deduced by the key param.
+       *
+       * @param key - The key to search for.
+       * @return A std::optional of the value corresponding to the key.
+       */
+      template <typename K>
+      std::optional<T> get(K&& key) {
```

Comment:
This seems somewhat error-prone, as there is nothing to enforce that Key has the correct type.  e.g. passing a literal 0 for a uint64_t index.

Reply:
I definitely agree. I haven't been able to come up with a solution yet though that allows us to template the kv_index on the key type in a way that can be deduced so that we don't have to specify it.

Reply:
Have you tried CTAD